core-primitives/attestation-handler/src/attestation_handler.rs

```diff
@@ -0,0 +1,648 @@
+		//       7. [out]p_qe_report need further check
+		//       8. [out]p_quote
+		//       9. quote_size
+
+		let spid: sgx_spid_t = Self::load_spid(RA_SPID_FILE)?;
+
+		let quote_result =
+			self.ocall_api.get_quote(sigrl_vec, report, sign_type, spid, quote_nonce)?;
+
+		let qe_report = quote_result.0;
+		let quote_content = quote_result.1;
+
+		// Added 09-28-2018
+		// Perform a check on qe_report to verify if the qe_report is valid
+		match rsgx_verify_report(&qe_report) {
+			Ok(()) => debug!("    [Enclave] rsgx_verify_report success!"),
+			Err(x) => {
+				error!("    [Enclave] rsgx_verify_report failed. {:?}", x);
+				return Err(x)
+			},
+		}
+
+		// Check if the qe_report is produced on the same platform
+		if target_info.mr_enclave.m != qe_report.body.mr_enclave.m
+			|| target_info.attributes.flags != qe_report.body.attributes.flags
+			|| target_info.attributes.xfrm != qe_report.body.attributes.xfrm
+		{
+			error!("    [Enclave] qe_report does not match current target_info!");
+			return Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+		}
+
+		debug!("    [Enclave] qe_report check success");
+
+		// Check qe_report to defend against replay attack
+		// The purpose of p_qe_report is for the ISV enclave to confirm the QUOTE
+		// it received is not modified by the untrusted SW stack, and not a replay.
+		// The implementation in QE is to generate a REPORT targeting the ISV
+		// enclave (target info from p_report) , with the lower 32Bytes in
+		// report.data = SHA256(p_nonce||p_quote). The ISV enclave can verify the
+		// p_qe_report and report.data to confirm the QUOTE has not be modified and
+		// is not a replay. It is optional.
+
+		// need to call this a second time (first time is when we get the sigrl revocation list)
+		// (has some internal state that needs to be reset)!
+		let ias_socket = self.ocall_api.get_ias_socket()?;
+
+		let mut rhs_vec: Vec<u8> = quote_nonce.rand.to_vec();
+		rhs_vec.extend(&quote_content);
+		let rhs_hash = rsgx_sha256_slice(&rhs_vec[..])?;
+		let lhs_hash = &qe_report.body.report_data.d[..32];
+
+		debug!("    [Enclave] rhs hash = {:02X}", rhs_hash.iter().format(""));
+		debug!("    [Enclave] lhs hash = {:02X}", lhs_hash.iter().format(""));
+
+		if rhs_hash != lhs_hash {
+			error!("    [Enclave] Quote is tampered!");
+			return Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+		}
+
+		let (attn_report, sig, cert) = self.get_report_from_intel(ias_socket, quote_content)?;
+		Ok((attn_report, sig, cert))
+	}
+
+	fn load_spid(filename: &str) -> SgxResult<sgx_spid_t> {
+		match io::read_to_string(filename).map(|contents| decode_spid(&contents)) {
+			Ok(r) => r,
+			Err(e) => {
+				error!("Failed to load SPID: {:?}", e);
+				Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+			},
+		}
+	}
+
+	fn get_ias_api_key() -> EnclaveResult<String> {
+		io::read_to_string(RA_API_KEY_FILE)
+			.map(|key| key.trim_end().to_owned())
+			.map_err(|e| EnclaveError::Other(e.into()))
+	}
+
+	pub fn create_ra_report_and_signature(
```

Comment:
If this function is not part of the trait, it probably doesn't have to be `pub`?

Reply:
No it is being called from outside, so it should be part of the trait. I will add it. 
The thing is that the trait is currently not really used. I tried to use it in the `global_components.rs` but quickly ran into lifetime issues. 
Do we use traits anywhere in the global components?

Reply:
The global components are all about the concrete types and instantiating them, so that's correct as it is. The trait will be used as soon as other components use the attestation handler.